server/src/main/java/org/opensearch/cluster/service/TaskBatcher.java

```diff
@@ -139,13 +142,13 @@ private void onTimeoutInternal(List<? extends BatchedTask> tasks, TimeValue time
             Object batchingKey = firstTask.batchingKey;
             assert tasks.stream().allMatch(t -> t.batchingKey == batchingKey)
                 : "tasks submitted in a batch should share the same batching key: " + tasks;
-            synchronized (tasksPerBatchingKey) {
-                LinkedHashSet<BatchedTask> existingTasks = tasksPerBatchingKey.get(batchingKey);
-                if (existingTasks != null) {
+            LinkedHashSet<BatchedTask> existingTasks = tasksPerBatchingKey.get(batchingKey);
```

Comment:
I'm concerned about a potential race with the following scenario:

Thread 1: executes line 98 and adds an empty LinkedHashSet to the map
Thread 2: executes lines 145-152, which will find an empty LinkedHashSet and remove it from the map
Thread 1: executes lines 103-118, which adds tasks to the LinkedHashSet but the set has been removed from the map

Is that scenario possible?

Reply:
+1 the race can possibly occur. The synchronize block on the existingTasks object is not sufficient to prevent any operation on that key in the main map ```tasksPerBatchingKey```.
@amkhar it is better to use ```computeIfPresent``` or  preferably```merge``` ( as it provides capability to provide default value as well) and pass the duplicate check etc in the remappingFunction as argument which will prevent any update to the same key from any other code path executing in different thread.

Reply:
@andrross @shwetathareja 
Sorry for delayed response.
Yes, race condition is possible. A little different than what Andrew explained. 

Thread 1 &2 both working on same batchingKey, and note that timeout method can only be called once thread has already executed line number 126 (only after adding the tasks).

Thread 1: executes till line number 126, so it has already added some list in the value of batchingKey
Thread 2: executes line number 98 and get a value from map which has a list of 1 task
Thread 1's async call back due to timeout: executes line 147-151 and removes the entry from map.
Thread 2: reaches on 103 and waits
Thread 2: executes the lines 104-117 and adds another entry in LinkedHashSet but it has already been removed from previous callback.

So, yes a fix is needed for this condition. I've found a way to handle this. It requires adding one line after 117 

```tasksPerBatchingKey.putIfAbsent(firstTask.batchingKey, existingTasks);```  **to make sure that if key-value was removed, we put it again while adding new tasks in the queue**. And we also need to add a check to remove already processed tasks like below at line number 116, because `runIfNotProcessed` method can also update and process some tasks in another thread.

```
if (existing.processed.get()) {
     tasksIterator.remove();
}
```

Something like this

```
synchronized (existingTasks) {
                Iterator<BatchedTask> tasksIterator = existingTasks.iterator();
                while (tasksIterator.hasNext()) {
                    BatchedTask existing = tasksIterator.next();
                    // check that there won't be two tasks with the same identity for the same batching key
                    BatchedTask duplicateTask = tasksIdentity.get(existing.getTask());
                    if (duplicateTask != null) {
                        throw new IllegalStateException(
                            "task ["
                                + duplicateTask.describeTasks(Collections.singletonList(existing))
                                + "] with source ["
                                + duplicateTask.source
                                + "] is already queued"
                        );
                    }
                    if (existing.processed.get()){
                        tasksIterator.remove();
                    }
                }
                existingTasks.addAll(tasks);
                tasksPerBatchingKey.putIfAbsent(firstTask.batchingKey, existingTasks);
            }
```

But thanks Shweta for suggesting `merge`, let me try that also. Looks similar to what we're trying to achieve.

Reply:
Thanks @amkhar I have one concern with the above proposed fix. The synchronized block is on existingTasks and ```putIfAbsent``` will add the key back again (in case it was removed during the timeout code execution form line 151). But lets says another Thread 3 came and at that time there was no entry in the map, it would have taken lock on a diff existingTasks object altogether. So now, 2 threads are executing the code which in synchronized block from line 103.

Reply:
@andrross 
@shwetathareja Agreed. That was another case. I realized that, taking separate locks on LinkedHashSet is problematic. If we just maintain the CHM bucket level lock, by using its library method, that will be clean and correct.

This is how code will look using **merge or computeIfPresent** (only writing the concerning code block)

**submitTasks** - create new or update existing list then update Map
```
LinkedHashSet<BatchedTask> newTasks = new LinkedHashSet<>(tasks);
tasksPerBatchingKey.merge(firstTask.batchingKey, newTasks, (existingTasks, updatedTasks) -> {
                for (BatchedTask existing : existingTasks) {
                    // check that there won't be two tasks with the same identity for the same batching key
                    BatchedTask duplicateTask = tasksIdentity.get(existing.getTask());
                    if (duplicateTask != null) {
                        throw new IllegalStateException(
                            "task ["
                                + duplicateTask.describeTasks(Collections.singletonList(existing))
                                + "] with source ["
                                + duplicateTask.source
                                + "] is already queued"
                        );
                    }
                }
                existingTasks.addAll(updatedTasks);
                return existingTasks;
            });
```

And **runIfNotProcessed** - here for iteration lock is not needed, because no other flow will be accessing the same object after it's removed from the Map.

```
LinkedHashSet<BatchedTask> pending = tasksPerBatchingKey.remove(updateTask.batchingKey);
            if (pending != null) {
                for (BatchedTask task : pending) {
                    if (task.processed.getAndSet(true) == false) {
                        logger.trace("will process {}", task);
                        toExecute.add(task);
                        processTasksBySource.computeIfAbsent(task.source, s -> new ArrayList<>()).add(task);
                    } else {
                        logger.trace("skipping {}, already processed", task);
                    }
                }
            }
```

**onTimeoutInterval** - Re-compute the value once timed out tasks are removed, and remove it if nothing left in LinkedHashSet.

```
tasksPerBatchingKey.computeIfPresent(batchingKey, (tasksKey, currentTasks) -> {
                currentTasks.removeAll(toRemove);
                if (currentTasks.isEmpty()) {
                    return null;
                }
                return currentTasks;
            });
```